### PR TITLE
Custom drawable layer v2

### PR DESCRIPTION
**Pull request by stefankarschti**
Introduce a generic interface for developing Custom Drawable Layers at core C++ level.
Issue #1735

### CODE REVIEW

**Comment by stefankarschti** [comment]
Simulator Screen Recording - iPad (10th generation) - 2023-10-23 at 22 57 27

**Review by sjg-wdw — APPROVED** [review]
Looks pretty good!

**Review by louwers — APPROVED** [review]

**Comment by louwers** [comment]
Bloaty says +0.5%.

Should we make it opt-in since you need to compile the library yourself to use this anyway?

File: include/mbgl/style/layers/custom_drawable_layer.hpp [hunk 1]
```diff
@@ -6,30 +6,101 @@
...
+
+    /**
+     * @brief Set the Tile ID
+     *
+     * @param tileID
+     */
+    void setTileID(OverscaledTileID tileID);
+
+    void setColor(Color color);
+    void setBlur(float blur);
```
[the remark this answers is not in the document]
Comment (stefankarschti) [reply include/mbgl/style/layers/custom_drawable_layer.hpp:58]: Polylines and other primitives could share some properties ([example: Fill picks up Color and Opacity](https://github.com/stefankarschti/maplibre-native/blob/a6876cad7dabdeef7c6e5598b852e62d43b5824e/platform/darwin/app/ExampleCustomDrawableStyleLayer.mm#L52C1-L132C6))

I agree that I clear mapping should be done, if a property is only applicable to lines we should call it `*LineProperty', for example.

**Review by TimSylvester — APPROVED** [review]

File: include/mbgl/style/layers/custom_drawable_layer.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefankarschti) [reply include/mbgl/style/layers/custom_drawable_layer.hpp]: in regards to "mode", the interface will allow mixing different primitive types transparently. new drawables will be generated and tweakers attached when primitive type or properties change.

**Review by mwilsnd — APPROVED** [review]

**Comment by stefankarschti** [comment]
@louwers 
> Bloaty says +0.5%.
> 
> Should we make it opt-in since you need to compile the library yourself to use this anyway?
> 
> Edit: Quoting Marc
> 
> > Probably included the first reference to it somehow via your PR

looks like the size impact is smaller after all.
```
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  +0.1%  +130Ki  +0.1% +6.87Ki    TOTAL
```

also. recompilation is not needed, just include the headers and link with the core library so that `CustomDrawableLayer` and `CustomDrawableLayerHost` are linked.

**Pull request merged**